SpatialGDK/Source/SpatialGDKServices/Private/SpatialCommandUtils.cpp

```diff
@@ -496,4 +497,55 @@ bool SpatialCommandUtils::GetProcessInfoFromPort(int32 Port, FString& OutPid, FS
 	return false;
 }
 
+void SpatialCommandUtils::FetchRuntimeBinary(const FString& RuntimeVersion)
+{
+	FString RuntimePath = FPaths::Combine(SpatialGDKServicesConstants::GDKProgramPath, TEXT("runtime"), RuntimeVersion);
+	SpatialCommandUtils::FetchPackageBinary(RuntimeVersion, SpatialGDKServicesConstants::RuntimeExe, "runtime", RuntimePath, true);
+}
+
+void SpatialCommandUtils::FetchInspectorBinary(const FString& InspectorVersion)
+{
+	FString InspectorPath = FPaths::Combine(SpatialGDKServicesConstants::GDKProgramPath, TEXT("inspector"), InspectorVersion, SpatialGDKServicesConstants::InspectorExe);
+	SpatialCommandUtils::FetchPackageBinary(InspectorVersion, SpatialGDKServicesConstants::InspectorExe, "inspector", InspectorPath, false);
+}
+
+void SpatialCommandUtils::FetchPackageBinary(const FString& PackageVersion, const FString& PackageExe, const FString& PackageName, const FString& SaveLocation, const bool bUnzip)
+{
+	FString PackagePath = FPaths::Combine(SpatialGDKServicesConstants::GDKProgramPath, *PackageName, PackageVersion);
+
+	// Check if the binary already exists for a given version
+	if (FPaths::FileExists(FPaths::Combine(PackagePath, PackageExe)))
+	{
+		UE_LOG(LogSpatialCommandUtils, Verbose, TEXT("%s binary already exists."), *PackageName);
+		return;
+	}
+
+	// If it does not exist then fetch the binary using `spatial worker package retrieve`
+	UE_LOG(LogSpatialCommandUtils, Log, TEXT("Trying to fetch %s version %s"), *PackageName, *PackageVersion);
+	FString Params = FString::Printf(TEXT("package retrieve %s %s %s %s"), *PackageName, *SpatialGDKServicesConstants::PlatformVersion,
+									 *PackageVersion, *SaveLocation);
+	if (bUnzip)
+	{
+		Params += " --unzip";
+	}
+
+	TOptional<FMonitoredProcess> FetchingProcess;
+	const FString& ExePath = SpatialGDKServicesConstants::SpatialExe;
+	FetchingProcess = { ExePath, Params, true, true };
+	FetchingProcess->OnOutput().BindLambda([](const FString& Output) {
+		UE_LOG(LogSpatialCommandUtils, Display, TEXT("Runtime: %s"), *Output);
+	});
+	FetchingProcess->Launch();
+
+	while (FetchingProcess->Update())
```

Comment:
I don't know exactly how Update() works, but as @MatthewSandfordImprobable  pointed out to me: 
this could potentially hang the editor? Or is it fine and the editor stays responsive?

Reply:
This call will eventually be wrapped in a AsyncTask so it can be blocking for now

Reply:
hmmmmm this is assuming that it will always be in an async task right? 
How about just using a timer and check if it's still runing after some time has passed?
Also now thinking more about it. Does the spatial CLI not have a time out themselves? do we want to add an additional time out on top of this?

Reply:
Aye - just as the other functions in this class are blocking when using `ExecuteAndReadOutput`. 
Want it to be blocking so it can return a bool rather than requiring a delegate to be passed for calling upon completion.

Regarding using a timer. I believe this design is more elegant and will return the instant the process returns. 
The reason I wanted us to use a `FMonitoredProcess` here was to also allow printing of the process's stdout/stderr while it's still running (compared to `ExectuteAndReadOutput` which will wait for an exit code before printing anything) and it's simpler to use vs `CreateProc` for this purpose. 

As to whether we need a timeout, spatial might have a timeout (most likely does) but it's not something we can control. Since this is part of the editor flow for actually starting the runtime I wanted something more robust for providing GUI feedback. If spatial hung for 10mins downloading the runtime then the editor would just sit there waiting for that time and the user could get frustrated. I think it's reasonable to say if it's not done after 1 min then something is amiss. We don't have this potential UX issue with `Setup.bat` downloading packages as that's ran outside the editor and part of setup.

Reply:
Generally, I would disagree with having a completely blocking editor function is more elegant. There are ways around this (e.g. adding a sleep, making an update function that ticks, timer, .. etc). However, I'm happy to revisit this in the final PR when we see all the different pieces together. :)

Regarding the timeout: The main concern I would have is if somebody increases the timeout here and then hits spatial's timeout. how big is the file? can we guarantee it will download it one minute? What if someone's internet connection is just really bad and they need more time. I would be cautious with making such assumptions as we have run into these problems before.

Reply:
Chatted offline. We'll catch up about the timeout in the final PR (after some investigation into the spatial toolbelt timeouts). Good for now